File: 2023/Java/Day1.java
// Java 21.0.1

import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;

public class Day1 {
    public static void main(String[] args) {
        String filePath = "../inputs/day1.txt";
        int total = 0;

        try {
            FileReader fileReader = new FileReader(filePath);
            BufferedReader bufferedReader = new BufferedReader(fileReader);
            String line;

            while ((line = bufferedReader.readLine()) != null) {
                System.out.println(line);
                char[] charArray = line.toCharArray();
                ArrayList<Character> numbers = new ArrayList<>();

                for (char character : charArray) {
                    if (Character.isDigit(character)) {
                        numbers.add(character);
                    }
                }

                System.out.println("numbers size = " + numbers.size());

                if (numbers.size() >= 2) {
                    String numeric = String.format("%c%c", numbers.get(0), numbers.get(numbers.size() - 1));
                    int number = Integer.parseInt(numeric);
                    total += number;
                } else if (numbers.size() == 1) {
                    String numeric = String.format("%c%c", numbers.get(0), numbers.get(0));
                    int number = Integer.parseInt(numeric);
                    total += number;
                }
            }

            System.out.println(total);
            
            fileReader.close();
            bufferedReader.close();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}

File: 2023/Scala/day1.scala
// Scala 3.3.1

import scala.collection.mutable.ArrayBuffer
import scala.io.Source


@main def day1() =
    val filePath: String = "../inputs/day1.txt"
    val source: Source = Source.fromFile(filePath)
    var total: Int = 0
    var count: Int = 0

    for (line <- source.getLines) {
        var numbers: ArrayBuffer[Char] = ArrayBuffer()

        for (character <- line) {
            if (character.isDigit) {
                numbers += character
            }
        }

        if (numbers.length >= 2) {
            val numeric: String = s"${numbers(0)}${numbers(numbers.length - 1)}"
            val number: Int = numeric.toInt
            total += number
        } else if (numbers.length == 1) {
            val numeric: String = s"${numbers(0)}${numbers(0)}"
            val number: Int = numeric.toInt
            total += number
        }

        count += 1
    }

    println(s"total = $total")

    source.close()